Attribute Mapping - Sync Frontend

puntope opened the pull request:
### Changes proposed in this Pull Request:

Closes part of #1648 .

This PR implements the Frontend par of https://github.com/woocommerce/google-listings-and-ads/pull/1731

That is to Poll the sync API endpoint and update the message in accordance. 

### Screenshots:

https://user-images.githubusercontent.com/5908855/196202508-c0345216-b5a1-4f93-9ef6-92dcd017cc6f.mov

### Detailed test instructions:
<!-- Add detailed instructions for how to test that this PR fixes the issue and confirm that it doesn't break any other features :) -->

1. Go to Attribute Mapping. See the message "Your products are synced with your rules"
2. Create a new rule and see how the message changes to "Your products are being synced with your rules" after a few seconds.
3. Without closing the window.  Go to WooCommerce - Status - Scheduler in another window an look for `gla/jobs/update_all_products/create_batch`. Scheduled for 30 minutes. Run it manually. 
4. Go back to the other window in Attribute mapping and see how after a few seconds it updates back to "Your products are synced with your rules"

puntope committed:
Integrate syncer frontend

eason9487 commented on js/src/hooks/usePolling.js:
```diff
@@ -0,0 +1,44 @@
+		if ( second === 0 && callCount > 0 && ( ! data || ! stopOnResolved ) ) {
+			start();
+		}
+	}, [ second, callCount, start, stopOnResolved, data ] );
```
As `data` is also in the deps array of this useEffect, if an API call responds within one second and changes the reference of `data`, it will invoke `start()` again which results in a duplicate API call.

https://user-images.githubusercontent.com/17420811/196396125-9c5a02f2-272a-4937-a78f-ebb8957eec07.mp4

eason9487 commented on js/src/hooks/usePolling.js:
```diff
@@ -0,0 +1,44 @@
+	delay = 5,
+	stopOnResolved = false
+) {
+	const { second, callCount, startCountdown } = useCountdown();
```
After leaving the Attribute Mapping page, it will cause an error `Cannot read properties of undefined (reading 'id')` in `useCountdown` and lead to a blank page. Considering `useCountdown` was changed to have the default `handle` value in the `develop` branch, we could cherry-pick f3cc0437ae9dd124aaac702c3b14099490e121ec into this PR or merge the `develop` branch to the shared branch of this project and rebase related PRs. So that the use case here could keep no parameters.

eason9487 reviewed (commented):
Found two issues that may need adjustments. Could you point me to the Figma document or mockup that this implementation is referred to? I couldn't find the match one from both design i1 and i2 documents on Figma.

puntope committed:
Merge branch 'feature/attribute-mapping' into add/attribute-mapping/syncer-frontend

# Conflicts:
#	tests/Unit/Jobs/UpdateAllProductsTest.php

puntope committed:
Fix usePolling duplicate request

puntope committed:
Merge branch 'add/attribute-mapping/syncer' into add/attribute-mapping/syncer-frontend

puntope commented:
Howdy @eason9487 

Thanks for cactus hing both bugs! 

I fixed them and this is ready for another round

eason9487 commented on js/src/hooks/usePolling.js:
```diff
@@ -0,0 +1,48 @@
+		[ fetch, startCountdown, delay ]
+	);
+
+	useEffect( () => {
+		if ( second === 0 && callCount > 0 ) {
+			start( ! data || ! stopOnResolved );
+		}
+	}, [ second, callCount, start, stopOnResolved ] );
```
There is an eslint warning `react-hooks/exhaustive-deps`. Unfortunately, the [issue of duplicate API calls](https://github.com/woocommerce/google-listings-and-ads/pull/1733#discussion_r997974904) still appears if adding `data` to useEffect deps array.

![image](https://user-images.githubusercontent.com/17420811/196844134-2366afeb-bca3-48eb-9b9c-94c990e19c2f.png)

With the revision [9003b36](https://github.com/woocommerce/google-listings-and-ads/blob/9003b36ca713b699db55f8eca7948942aaa095ab/js/src/hooks/usePolling.js#L34-L38), replacing `stopOnResolved` and `data` in deps with a combined variable such as `const shouldRun = ! data || ! stopOnResolved;`, which is declared out of the useEffect scope, could be a solution.

eason9487 reviewed (commented):
Thanks for updating this PR with `develop` branch. After some tests, I think the issue of duplicate API calls might not yet be solved. Furthermore, quoted from https://github.com/woocommerce/google-listings-and-ads/pull/1733#pullrequestreview-1145451070:
> Could you point me to the Figma document or mockup that this implementation is referred to? I couldn't find the match one from both design i1 and i2 documents on Figma.

I'm still not 100% sure which design document is the target one to be referred to. In fqR0EHi63lWahRcVTKCcba-fi-2305%3A162769, the most likely design documents I can find are the following. However, they are quite different from the actual implementation in this PR.

- Statuses
   ![image](https://user-images.githubusercontent.com/17420811/196847097-140e1fbe-4a75-4de3-b20a-66acf8d38d7b.png)
- Tooltip
   ![image](https://user-images.githubusercontent.com/17420811/196847146-cae07a86-3dc6-4cac-abbf-1f6e0ac926f0.png)

puntope committed:
Adapt sync frontend to new UX

puntope committed:
Merge branch 'add/attribute-mapping/syncer' into add/attribute-mapping/syncer-frontend

puntope committed:
Tweak Sync Component and tests

puntope commented:
Howdy @eason9487 

I updated the PR again with your awesome suggestions. 

Thanks for the tip in regards the useEffect problem

I updated the logic adapting the table to the new UX. Now it looks accurate (notice pagination in the center)

ℹ️  Since backend has changed after feedback, for getting the last sync date you will need to add this option in the Options Table `gla_update_all_products_last_sync`. You can add a timestamp like `1667338631` as value. Or just update an attribute rule and run manually the Job in Action Scheduler (or wait 30 mins) otherwise the date will be `never`. Notice also when there is an action scheduled or running, we don't show  the date but instead "Scheduled for sync"

puntope committed:
Temporarily Remove a problematic test

puntope committed:
Update test

eason9487 reviewed (approved):
Thanks for the updates. LGTM. There are a few styling differences that could be further adjusted but they are not blockers.

---

The ways of line break and alignment in the table footer are a little different from the design on Figma.
![image](https://user-images.githubusercontent.com/17420811/199400958-42abde66-1719-4505-811e-661f1dffc9f9.png)
![image](https://user-images.githubusercontent.com/17420811/199401043-4b0b1aa7-151f-4a76-a07a-0f80e3f4383b.png)

Figma:
![image](https://user-images.githubusercontent.com/17420811/199401394-46dda087-bb16-4fad-97d2-9d77001250cc.png)

puntope commented:
> The ways of line break and alignment in the table footer are a little different from the design on Figma

Thanks @eason9487 . This is noted for future tweaks
